File: pca_lite/search/semantic_scholar.py
"""Semantic Scholar API web search provider."""
import httpx
from typing import Any


class SemanticScholarSearch:
    """Search provider backed by Semantic Scholar Graph API."""

    BASE_URL = "https://api.semanticscholar.org/graph/v1"

    def __init__(self, api_key: str | None = None, timeout: int = 30):
        self.api_key = api_key
        self.timeout = timeout

    async def search(self, query: str, max_results: int = 10) -> list[dict[str, Any]]:
        """Search Semantic Scholar for papers.

        Args:
            query: Search query string.
            max_results: Maximum number of results.

        Returns:
            List of paper metadata dicts.
        """
        fields = (
            "title,authors,name,year,abstract,"
            "citationCount,openAccessPdf,externalIds,url,"
            "venue,publicationVenue"
        )
        params = {
            "query": query,
            "limit": min(max_results, 100),
            "fields": fields,
        }
        headers = {}
        if self.api_key:
            headers["x-api-key"] = self.api_key

        try:
            async with httpx.AsyncClient(timeout=self.timeout) as client:
                resp = await client.get(
                    f"{self.BASE_URL}/paper/search",
                    params=params,
                    headers=headers,
                )
                resp.raise_for_status()
                data = resp.json()

            papers = []
            for hit in data.get("data", []):
                paper = self._parse_paper(hit)
                papers.append(paper)
            return papers
        except Exception as e:
            print(f"[WARN] Semantic Scholar search failed: {e}")
            return []

    def _parse_paper(self, hit: dict) -> dict[str, Any]:
        author_names = [
            a.get("name", "Unknown") for a in hit.get("authors", [])
        ]
        external_ids = hit.get("externalIds", {}) or {}
        return {
            "title": hit.get("title", "Unknown"),
            "authors": author_names,
            "year": hit.get("year"),
            "abstract": hit.get("abstract", ""),
            "doi": external_ids.get("DOI", ""),
            "arxiv_id": external_ids.get("ArXiv", ""),
            "url": hit.get("url", ""),
            "open_access_pdf": (
                hit.get("openAccessPdf", {}).get("url", "")
                if hit.get("openAccessPdf")
                else ""
            ),
            "venue": hit.get("venue", ""),
            "citation_count": hit.get("citationCount", 0),
            "source": "semantic_scholar",
        }

File: crates/citeforge-search/src/semantic_scholar.rs
use async_trait::async_trait;
use citeforge_core::entity::LiteratureEntry;
use citeforge_core::ports::{SearchEngine, SearchError};

pub struct SemanticScholarClient {
    api_key: Option<String>,
    client: reqwest::Client,
}

impl SemanticScholarClient {
    pub fn new(api_key: Option<String>) -> Self {
        Self {
            api_key,
            client: reqwest::Client::new(),
        }
    }
}

#[async_trait]
impl SearchEngine for SemanticScholarClient {
    async fn search(&self, query: &str, max_results: usize) -> Result<Vec<LiteratureEntry>, SearchError> {
        let url = "https://api.semanticscholar.org/graph/v1/paper/search";

        let mut req = self.client.get(url)
            .query(&[
                ("query", query),
                ("limit", &max_results.to_string()),
                ("fields", "title,authors,abstract,year,venue,citationCount,externalIds"),
            ]);

        if let Some(key) = &self.api_key {
            req = req.header("x-api-key", key);
        }

        let resp = req.send().await
            .map_err(|e| SearchError::Network(e.to_string()))?;

        if resp.status() == 429 {
            return Err(SearchError::RateLimited);
        }

        if !resp.status().is_success() {
            return Err(SearchError::Api(format!("API error: {}", resp.status())));
        }

        let result: serde_json::Value = resp.json().await
            .map_err(|e| SearchError::Network(e.to_string()))?;

        if let Some(msg) = result["error"].as_str() {
            tracing::warn!("Semantic Scholar API error: {}", msg);
        }

        let entries: Vec<LiteratureEntry> = result["data"]
            .as_array()
            .unwrap_or(&vec![])
            .iter()
            .filter_map(|paper| {
                let paper_id = paper["paperId"].as_str()?;
                if paper_id.is_empty() {
                    return None;
                }
                Some(paper)
            })
            .enumerate()
            .map(|(idx, paper)| {
                let authors: Vec<String> = paper["authors"]
                    .as_array()
                    .unwrap_or(&vec![])
                    .iter()
                    .filter_map(|a| a["name"].as_str().map(|s| s.to_string()))
                    .collect();

                let year = paper["year"].as_i64().map(|y| y as i32);
                let venue = paper["venue"].as_str().map(|s| s.to_string());
                let citation_count = paper["citationCount"].as_i64().map(|c| c as i32);
                let doi = paper["externalIds"]["DOI"].as_str().map(|s| s.to_string());

                LiteratureEntry {
                    id: paper["paperId"].as_str().unwrap().to_string(),
                    sort_order: idx as i32 + 1,
                    title: paper["title"].as_str().unwrap_or("").to_string(),
                    authors,
                    abstract_text: paper["abstract"].as_str().map(|s| s.to_string()),
                    doi,
                    year,
                    venue,
                    citation_count,
                    verified: false,
                }
            })
            .collect();

        Ok(entries)
    }
}
